types/wlr_surface.c

```diff
@@ -437,6 +437,8 @@ static void surface_cache_pending(struct wlr_surface *surface) {
 	wl_list_insert(surface->cached.prev, &cached->cached_state_link);
 
 	surface->pending.seq++;
+
+	surface->last_committed = cached;
```

Comment:
Is this not problematic if a client commits two times and we have two cached events? Maybe the frame callback was in the first commit. This way we're losing the callbacks because the next commit is unrelated. IDK if real-world clients do this, but it is a possibility.

Reply:
Hmm, but we ony render the last state we got, so it would make sense to only call last commit callbacks, but that's just my opinion.

What does Weston do in this case?

Reply:
Weston seems to be accumulating all frames in a single list when committing. https://gitlab.freedesktop.org/wayland/weston/-/blob/main/libweston/compositor.c#L3859

Also, the wayland protocol indicates that this is the correct thing to do:

```
For a wl_surface, the notifications are posted in the order the frame requests were committed.
```

So to my understanding, we should accumulate.

Reply:
Okay, accumulate we will. Thanks for checking.